From d33e319f146fecbdfc0cab604a56b054a38fc389 Mon Sep 17 00:00:00 2001
From: Sungtak Lee <taklee@google.com>
Date: Wed, 25 Sep 2024 23:27:58 +0000
Subject: [PATCH 070/436] MediaMetadataRetriever: add static lock for image
 extraction

MediaMetadataRetriever binder service has a static lock in
MetadataRetrieverClient. The static lock serializes getFrameAtTime(),
getImageAtIndex(), getImageRectAtIndex() and getFrameAtIndex()
regardless of where the transaction being originated.

The static lock in the service side throttles image extraction not to
consume too much memory. But the static lock can cause to consume all
the binder thread pools and lead to starvation.

By adding a static lock to the client side(MediaMetadataRetriever),
image extraction from the same process will consume only one thread in
the thread pool.

Flag: EXEMPT bugfix
Bug: 359802236
Test: presubmit
Change-Id: I8476df2aac34506d7571b98e1c24aee8d74a7900
---
 media/libmedia/include/media/mediametadataretriever.h | 4 ++++
 media/libmedia/mediametadataretriever.cpp             | 6 ++++++
 2 files changed, 10 insertions(+)

diff --git a/media/libmedia/include/media/mediametadataretriever.h b/media/libmedia/include/media/mediametadataretriever.h
index 116ed9aa3f..d76ed25fc9 100644
--- a/media/libmedia/include/media/mediametadataretriever.h
+++ b/media/libmedia/include/media/mediametadataretriever.h
@@ -122,6 +122,10 @@ private:
     static sp<IMediaPlayerService>            sService;
 
     Mutex                                     mLock;
+    // Static lock was added to the client in order to consume at most
+    // one service thread from image extraction requests of the same
+    // client process(See also b/21277449).
+    static Mutex                              sLock;
     sp<IMediaMetadataRetriever>               mRetriever;
 
 };
diff --git a/media/libmedia/mediametadataretriever.cpp b/media/libmedia/mediametadataretriever.cpp
index 40fd022812..9196f9f97b 100644
--- a/media/libmedia/mediametadataretriever.cpp
+++ b/media/libmedia/mediametadataretriever.cpp
@@ -35,6 +35,8 @@ Mutex MediaMetadataRetriever::sServiceLock;
 sp<IMediaPlayerService> MediaMetadataRetriever::sService;
 sp<MediaMetadataRetriever::DeathNotifier> MediaMetadataRetriever::sDeathNotifier;
 
+Mutex MediaMetadataRetriever::sLock;
+
 const sp<IMediaPlayerService> MediaMetadataRetriever::getService()
 {
     Mutex::Autolock lock(sServiceLock);
@@ -143,6 +145,7 @@ sp<IMemory> MediaMetadataRetriever::getFrameAtTime(
     ALOGV("getFrameAtTime: time(%" PRId64 " us) option(%d) colorFormat(%d) metaOnly(%d)",
             timeUs, option, colorFormat, metaOnly);
     Mutex::Autolock _l(mLock);
+    Mutex::Autolock _gLock(sLock);
     if (mRetriever == 0) {
         ALOGE("retriever is not initialized");
         return NULL;
@@ -155,6 +158,7 @@ sp<IMemory> MediaMetadataRetriever::getImageAtIndex(
     ALOGV("getImageAtIndex: index(%d) colorFormat(%d) metaOnly(%d) thumbnail(%d)",
             index, colorFormat, metaOnly, thumbnail);
     Mutex::Autolock _l(mLock);
+    Mutex::Autolock _gLock(sLock);
     if (mRetriever == 0) {
         ALOGE("retriever is not initialized");
         return NULL;
@@ -167,6 +171,7 @@ sp<IMemory> MediaMetadataRetriever::getImageRectAtIndex(
     ALOGV("getImageRectAtIndex: index(%d) colorFormat(%d) rect {%d, %d, %d, %d}",
             index, colorFormat, left, top, right, bottom);
     Mutex::Autolock _l(mLock);
+    Mutex::Autolock _gLock(sLock);
     if (mRetriever == 0) {
         ALOGE("retriever is not initialized");
         return NULL;
@@ -180,6 +185,7 @@ sp<IMemory>  MediaMetadataRetriever::getFrameAtIndex(
     ALOGV("getFrameAtIndex: index(%d), colorFormat(%d) metaOnly(%d)",
             index, colorFormat, metaOnly);
     Mutex::Autolock _l(mLock);
+    Mutex::Autolock _gLock(sLock);
     if (mRetriever == 0) {
         ALOGE("retriever is not initialized");
         return NULL;
-- 
2.43.0

